crypto/thread/api.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2019-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#if defined(OPENSSL_THREADS)
+
+# include <openssl/crypto.h>
+# include <internal/worker.h>
+# include <internal/thread.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+int CRYPTO_THREAD_enabled(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    return tdata->enabled;
+}
+
+int CRYPTO_THREAD_enable(OPENSSL_CTX *ctx, int max_threads)
+{
+    OPENSSL_CTX_THREADS tdata;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    tdata->threads.cap = max_threads;
+    tdata->enabled = 1;
+    return 1;
+}
+
+int CRYPTO_THREAD_disable(OPENSSL_CTX *ctx)
+{
+    OPENSSL_CTX_THREADS tdata;
+    CRYPTO_WORKER worker;
+    struct list *iter, *tmp;
+
+    tdata = OPENSSL_CTX_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->threads.cap = 0;
+    tdata->enabled = 0;
+
+    list_for_each_safe(iter, tmp, &tdata->workers.available) {
```

Comment:
is that macro portable to posix?

Reply:
The macro `container_of` is defined as:
```# define container_of(ptr, type, member) ((type *)((char *)ptr - offsetof(type, member)))``` (alternative definition can be found here: https://en.wikipedia.org/wiki/Offsetof). The `list*` macros are simply for statements wrapped for readability purposes. As for `offsetof` that `container_of` uses: https://man7.org/linux/man-pages/man3/offsetof.3.html